File: 백준/Gold/1300. K번째 수/K번째 수.py
import sys
N = int(sys.stdin.readline())
K = int(sys.stdin.readline())
start = 1
end = K

while start <= end:
    mid = (start + end) // 2
    cnt = 0
    # ex) N*N에서 20보다 작거나 같은 수는
    # 1행 기준 1~ 20 -> 20//1 개
    # 2행 기준 2,4,6,8... -> 20//2 개
    # 3행 기준 3,6,9,12... -> 20//3 개
    # 4행 기준 4,8,12,16,20... -> 20//4개
    # N행 기준 N,N*2,N*3... -> 20//N개!
    # 이중 찾는값//i가 N보다 큰 경우가 존재하므로 최대 N까지만 갖게 설정
    for i in range(1, N+1):
        cnt += min(mid//i, N)
    if cnt >= K:
        result = mid
        end = mid - 1
    else:
        start = mid + 1
print(result)

File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static int n,m;
    public static int result = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        m = Integer.parseInt(br.readLine());

        // 1. n*n했을 때 나올 수 있는 전체 숫자를 이분탐색으로 구한다.
        int s = 1;
        int e = m;
        int mid;

        while(s <= e){
            mid = (s + e) / 2;

            // 중간에 있는 값(t) t이하 숫자의 개수가 몇 개인지 구하기
            int count = find_under(mid);

            if(count >= m){
                e = mid - 1;
                result = mid;
            }else{
                s = mid + 1;
            }
        }


        System.out.println(result);
    }
    public static int find_under(int num){
        int total = 0;
        for(int i = 1 ; i <= n ; i++){
            total += Math.min(n, num / i);
        }

        return total;
    }
}